src/freertos_drivers/spiffs/SPIFFS.cxx

```diff
@@ -0,0 +1,537 @@
+{
+    spiffs_file fd = file->privInt;
+
+    ssize_t result = SPIFFS_read(&fs_, fd, buf, count);
+
+    if (result < 0)
+    {
+        return -errno_translate(result);
+    }
+
+    return result;
+}
+
+//
+// SPIFFS::write()
+//
+ssize_t SPIFFS::write(File *file, const void *buf, size_t count)
+{
+    spiffs_file fd = file->privInt;
+
+    ssize_t result = SPIFFS_write(&fs_, fd, (void*)buf, count);
+
+    if (result < 0)
+    {
+        return -errno_translate(result);
+    }
+
+    return result;
+}
+
+///
+// SPIFFS::lseek()
+//
+off_t SPIFFS::lseek(File* file, off_t offset, int whence)
+{
+    spiffs_file fd = file->privInt;
+    int spiffs_whence;
+
+    switch (whence)
+    {
+        default:
+            return (off_t)-EINVAL;
+        case SEEK_SET:
+            spiffs_whence = SPIFFS_SEEK_SET;
+            break;
+        case SEEK_CUR:
+            spiffs_whence = SPIFFS_SEEK_CUR;
+            break;
+        case SEEK_END:
+            spiffs_whence = SPIFFS_SEEK_END;
+            break;
+    }
+
+    off_t result = SPIFFS_lseek(&fs_, fd, offset, spiffs_whence);
+
+    if (result < 0)
+    {
+        return -errno_translate(result);
+    }
+
+    return result;
+}
+
+//
+// SPIFFS::fstat()
+//
+int SPIFFS::fstat(File* file, struct stat *stat)
+{
+    spiffs_file fd = file->privInt;
+    spiffs_stat ffs_stat;
+
+    ssize_t result = SPIFFS_fstat(&fs_, fd, &ffs_stat);
+
+    if (result < 0)
+    {
+        return -errno_translate(result);
+    }
+
+    memset(stat, 0, sizeof(*stat));
+    stat->st_ino = ffs_stat.obj_id;
+    stat->st_size = ffs_stat.size;
+    switch (ffs_stat.type)
+    {
+        default:
+            break;
+        case SPIFFS_TYPE_FILE:
+            stat->st_mode = S_IFREG;
+            break;
+        case SPIFFS_TYPE_DIR:
+            stat->st_mode = S_IFDIR | S_IXUSR | S_IXGRP | S_IXOTH;
+            break;
+        case SPIFFS_TYPE_HARD_LINK:
+            stat->st_mode = S_IFLNK | S_IXUSR | S_IXGRP | S_IXOTH;
+            break;
+        case SPIFFS_TYPE_SOFT_LINK:
+            stat->st_mode = S_IFLNK | S_IXUSR | S_IXGRP | S_IXOTH;
+            break;
+    }
+
+    stat->st_mode |= S_IRUSR | S_IRGRP | S_IROTH |
+                     S_IWUSR | S_IWGRP | S_IWOTH;
+
+    return 0;
+}
+
+//
+// SPIFFS::stat()
+//
+int SPIFFS::stat(const char *path, struct stat *stat)
+{
+    spiffs_stat ffs_stat;
+
+    if (!strcmp(path, "") || !strcmp(path, "/"))
+    {
+        // this is the root directory, which cannot be stat
+        ffs_stat.type = SPIFFS_TYPE_DIR;
+        ffs_stat.obj_id = 0;
+        ffs_stat.size = 0;
+    }
+    else
+    {
+        ssize_t result = SPIFFS_stat(&fs_, path, &ffs_stat);
+
+        if (result < 0)
+        {
+            return -errno_translate(result);
+        }
+    }
+
+    memset(stat, 0, sizeof(*stat));
+    stat->st_ino = ffs_stat.obj_id;
+    stat->st_size = ffs_stat.size;
+    switch (ffs_stat.type)
+    {
+        default:
+            break;
+        case SPIFFS_TYPE_FILE:
```

Comment:
consider refacroring the shared code between the two stat functions to a common helper function.

Reply:
Done.